pos_accounting/src/pool/storage/mod.rs

```diff
@@ -20,7 +28,169 @@ impl<'a, S> PoSAccountingDBMut<'a, S> {
     }
 }
 
+pub struct DataMergeUndo {
+    pool_data_undo: BTreeMap<H256, Option<PoolData>>,
+    delegation_data_undo: BTreeMap<H256, Option<DelegationData>>,
+}
+
 impl<'a, S: PoSAccountingStorageWrite> PoSAccountingDBMut<'a, S> {
+    pub fn merge_with_delta(
+        &mut self,
+        other: PoSAccountingDeltaData,
+    ) -> Result<DataMergeUndo, Error> {
+        let pool_data_undo = self.merge_data_generic(
```

Comment:
I'm very happy we managed to get this to work. See how beautiful and clean it's!